chia/full_node/block_height_map.py

```diff
@@ -0,0 +1,204 @@
+import aiosqlite
+import logging
+from typing import Dict, List, Optional, Tuple
+from chia.util.ints import uint32
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.types.blockchain_format.sub_epoch_summary import SubEpochSummary
+from pathlib import Path
+import aiofiles
+from dataclasses import dataclass
+from chia.util.streamable import Streamable, streamable
+
+log = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+@streamable
+class SesCache(Streamable):
+    content: List[Tuple[uint32, bytes]]
+
+
+class BlockHeightMap:
+    db: aiosqlite.Connection
+
+    # the below dictionaries are loaded from the database, from the peak
+    # and back in time on startup.
+
+    # Defines the path from genesis to the peak, no orphan blocks
+    # this buffer contains all block hashes that are part of the current peak
+    # ordered by height. i.e. __height_to_hash[0..32] is the genesis hash
+    # __height_to_hash[32..64] is the hash for height 1 and so on
+    __height_to_hash: bytearray
+
+    # All sub-epoch summaries that have been included in the blockchain from the beginning until and including the peak
+    # (height_included, SubEpochSummary). Note: ONLY for the blocks in the path to the peak
+    # The value is a serialized SubEpochSummary object
+    __sub_epoch_summaries: Dict[uint32, bytes]
+
+    # count how many blocks have been added since the cache was last written to
+    # disk
+    __dirty: int
+
+    # the file we're saving the height-to-hash cache to
+    __height_to_hash_filename: Path
+
+    # the file we're saving the sub epoch summary cache to
+    __ses_filename: Path
+
+    @classmethod
+    async def create(cls, blockchain_dir: Path, db: aiosqlite.Connection) -> "BlockHeightMap":
+        self = BlockHeightMap()
+        self.db = db
+
+        self.__dirty = 0
+        self.__height_to_hash = bytearray()
+        self.__sub_epoch_summaries = {}
+        self.__height_to_hash_filename = blockchain_dir / "height-to-hash"
+        self.__ses_filename = blockchain_dir / "sub-epoch-summaries"
+
+        res = await self.db.execute(
+            "SELECT header_hash,prev_hash,height,sub_epoch_summary from block_records WHERE is_peak=1"
+        )
+        row = await res.fetchone()
+        await res.close()
+
+        if row is None:
+            return self
+
+        try:
+            async with aiofiles.open(self.__height_to_hash_filename, "rb") as f:
+                self.__height_to_hash = bytearray(await f.read())
+        except Exception:
+            # it's OK if this file doesn't exist, we can rebuild it
+            pass
+
+        try:
+            async with aiofiles.open(self.__ses_filename, "rb") as f:
+                self.__sub_epoch_summaries = {k: v for (k, v) in SesCache.from_bytes(await f.read()).content}
```

Comment:
this is because `Streamable` doesn't support dictionaries...

Reply:
Is it possible that the two files fall out of sync? What would happen in that case?

Reply:
Should we load the blocks again if the SES file is wrong?

Reply:
it is possible (under extraordinary circumstances). This is taken into account when they are loaded. We sync with the database until we find a match in *both* maps.

This is covered by tests as well.